basic/airdrop-with-withdraw/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+use scrypto::prelude::*;
+
+
+blueprint! {
+
+    struct AirdropWithWithdraw {
+        admin_badge: Address,
+        recipients: HashMap<Address, Vault>,
+        recipient_badge_def : ResourceDef,
+        minter_badge_vault : Vault
+    }
+
+    impl AirdropWithWithdraw {
+        
+        pub fn new() -> (Component, Bucket) {
+
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                .initial_supply_fungible(1);
+
+            let minter_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                               .metadata("name", "minter badge")
+                                               .initial_supply_fungible(1); 
+
+            let recipient_badge_def = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                .flags(MINTABLE | INDIVIDUAL_METADATA_MUTABLE)
+                                .badge(minter_badge.resource_address(), MAY_MINT | MAY_TRANSFER | MAY_CHANGE_INDIVIDUAL_METADATA)
+                                .no_initial_supply(); 
+            
+            let component = Self {
+                admin_badge : admin_badge.resource_address(),
+                recipients :  HashMap::new(),
+                recipient_badge_def ,
+                minter_badge_vault : Vault::with_bucket(minter_badge)
+            }
+            .instantiate();
+
+            return (component, admin_badge)
+        }
+
+
+        #[auth(admin_badge)]
+        pub fn add_recipient(&mut self, recipient: Address, tokens : Bucket)  {
+            assert!(tokens.amount() > Decimal::zero(), "tokens quantity cannot be 0"); 
+            assert!(!self.recipients.contains_key(&recipient) , "Recipient already exist");
+            self.recipients.insert(recipient, Vault::with_bucket(tokens));
+            let user_badge = self.minter_badge_vault.authorize(|auth|
+                return self.recipient_badge_def.mint(1, auth) 
+            );
+            Account::from(recipient).deposit(user_badge);
+        }
+        
+
+        #[auth(recipient_badge_def)]
+        pub fn withdraw_token(&mut self) -> Bucket
```

Comment:
I am not sure if this is what is causing the issue, but when you call add_recipient, the key you insert into the recipients map is the address of the account. Then in the withdraw_token, you try to find the entry in the map by specifying the address of the badge as key which does not exists.

I just realised that you will need to create a non fungible token instead of a badge for user identification, since every badges is the same so every user will be withdrawing from the same vault. Each NFT have a specific ID which can be used as the key in the recipients map.

I suggest looking at my example on how to create NFTs: https://github.com/cbisaillon/Scrypto-Advent-Calendar/blob/master/day17/src/lib.rs. Notice how the ID of the NFT is just the number of NFT created. That way you are certain that it will always be unique

Reply:
Hello Clement, 
Thanks for the review. Indeed, I used account  address  as a key when insert into HashMap and when retrieving the address of the badge. I just updated the implementation with non fungible token  but I still have the Insufficient balance problem.

Reply:
The problem is the line `let recipient_badge = receipt.resource_def(1).unwrap(); `. This references the minter badge, not the recipient's badge. The index should be 2.

Reply:
Thanks for the review, I just updated the implementation and it work fine now.